testing/handlertest/testhandler.go

```diff
@@ -28,67 +33,65 @@ import (
 //		return context.WithValue(testCtx, requestStateKey{}, &requestState{r: r, w: w}), features
 //	}
 //
-//	if err := handlertest.TestHost(host{}, newCtx); err != nil {
-//		t.Fatal(err)
-//	}
-func TestHost(h handler.Host, newCtx func(handler.Features) (context.Context, handler.Features)) error {
-	t := hostTester{h: h, newCtx: newCtx}
-
-	t.checkMethod()
-	t.checkURI()
-	t.checkProtocolVersion()
-	t.checkRequestHeaders()
-	t.checkRequestBody()
-	t.checkRequestTrailers()
-	t.checkStatusCode()
-	t.checkResponseHeaders()
-	t.checkResponseBody()
-	t.checkResponseTrailers()
-
-	if len(t.errText) == 0 {
+//	// Run all the tests
+//	handlertest.HostTest(t, host{}, newCtx)
+func HostTest(t *testing.T, h handler.Host, newCtx func(handler.Features) (context.Context, handler.Features)) error {
+	ht := hostTester{t: t, h: h, newCtx: newCtx}
+
+	ht.testMethod()
+	ht.testURI()
+	ht.testProtocolVersion()
+	ht.testRequestHeaders()
+	ht.testRequestBody()
+	ht.testRequestTrailers()
+	ht.testStatusCode()
+	ht.testResponseHeaders()
+	ht.testResponseBody()
+	ht.testResponseTrailers()
+
+	if len(ht.errText) == 0 {
 		return nil
 	}
-	return errors.New("TestHost found errors:\n" + string(t.errText))
+	return errors.New("TestHost found errors:\n" + string(ht.errText))
 }
 
 // A hostTester holds state for running the test.
 type hostTester struct {
+	t       *testing.T
 	h       handler.Host
 	newCtx  func(handler.Features) (context.Context, handler.Features)
 	errText []byte
 }
 
-// errorf adds an error line to errText.
-func (t *hostTester) errorf(format string, args ...any) {
-	if len(t.errText) > 0 {
-		t.errText = append(t.errText, '\n')
-	}
-	t.errText = append(t.errText, fmt.Sprintf(format, args...)...)
-}
-
-func (t *hostTester) checkMethod() {
-	ctx, _ := t.newCtx(0) // no features required
+func (h *hostTester) testMethod() {
+	ctx, _ := h.newCtx(0) // no features required
 
-	// Check default
-	if want, have := "GET", t.h.GetMethod(ctx); want != have {
-		t.errorf("GetMethod: unexpected default, want: %v, have: %v", want, have)
-	}
+	h.t.Run("GetMethod default", func(t *testing.T) {
+		// Check default
+		if want, have := "GET", h.h.GetMethod(ctx); want != have {
+			t.Errorf("unexpected default method, want: %v, have: %v", want, have)
+		}
+	})
 
-	for _, want := range []string{"POST", "OPTIONS"} {
-		t.h.SetMethod(ctx, want)
+	h.t.Run("SetMethod", func(t *testing.T) {
```

Comment:
note, not doing t.Run for every dimension as it can quickly get out of hand, especially after we migrate in more tests. Instead the error format will say which failed.